### PR TITLE
refactor(dataset): dataset meets datastore

### PR DESCRIPTION
**Pull request by tianweidut**
## Description
- issue: #896 
- refactor dataset build and loader with datastore
- skip tests for job create or cmp, wait for the pr: https://github.com/star-whale/starwhale/pull/880 

## Feature Show
- dataset build
image
- dataset loader
image
image

## Modules
- [x] Client
- [x] Python-SDK

## Checklist
- [x] run code format and lint check
- [x] add unit test
- [ ] add necessary doc

### CODE REVIEW

File: client/starwhale/api/_impl/dataset.py [hunk 1]
```diff
@@ -24,6 +36,132 @@
...
+
+
+class ObjectStoreType:
+    LOCAL = "l"
+    REMOTE = "r"
+
+
+class DataOriginType:
+    NEW = "n"
+    INHERIT = "i"
```
Comment (jialeicui) [inline client/starwhale/api/_impl/dataset.py:51]: Consider subclass of `Enum`?

File: client/starwhale/api/_impl/dataset.py [hunk 2]
```diff
@@ -24,6 +36,132 @@
...
+    def _do_validate(self) -> None:
+        _ok, _reason = validate_obj_name(self.name)
+        if not _ok:
+            raise InvalidObjectName(f"{self.name}: {_reason}")
+
+        if not self.version:
+            raise FieldTypeOrValueError("no version field")
+
+    def __str__(self) -> str:
+        return f"Dataset Table: {self._ds_wrapper._meta_table_name}"
```
Comment (jialeicui) [inline client/starwhale/api/_impl/dataset.py:130]: Make `_meta_table_name` an accessible func or prop?

File: client/starwhale/api/_impl/dataset.py [hunk 3]
```diff
@@ -24,6 +36,132 @@
...
+
+    def put(self, row: TabularDatasetRow) -> None:
+        self._ds_wrapper.put(row.id, **row.asdict())
+
+    def scan_all(self) -> t.Generator[TabularDatasetRow, None, None]:
+        return self.scan(self.start, self.end)
+
+    def scan(self, start: int, end: int) -> t.Generator[TabularDatasetRow, None, None]:
+        for _d in self._ds_wrapper.scan(start, end):
+            yield TabularDatasetRow(**_d)
```
Comment (jialeicui) [inline client/starwhale/api/_impl/dataset.py:146]: When a function accepts a  `TabularDataset` with a predefined range (e.g. called `set_range(2, 4)` outside).
Calling dataset.scan_all() will get [2, 3], and dataset.scan(0, 2) will get [0, 1].
It doesn't feel as expected.

Comment (tianweidut) [reply client/starwhale/api/_impl/dataset.py:146]: fixed

File: client/starwhale/api/_impl/dataset.py (hunk 2 above)
Comment (tianweidut) [reply client/starwhale/api/_impl/dataset.py]: updated

File: client/starwhale/api/_impl/dataset.py (hunk 1 above)
Comment (tianweidut) [reply client/starwhale/api/_impl/dataset.py:55]: use `Enum` to refactor.

**Comment by jialeicui** [comment]
LGTM

**Review by jialeicui — APPROVED** [review]

**Pull request merged**